### ISSUE TITLE
Use goog.structs.LinkedMap to implement the cljs.core/IFn protocol, But got a TypeError message

### ISSUE THREAD

**Issue by DevinKin** [thheller/shadow-cljs#742]
### Version Info
shadow-cljs 2.10.8
java 8
windows 10
***
### Sample Code
here is my code

```clojure
(ns app.main
  (:require [goog.structs.LinkedMap]))
(extend-type goog.structs.LinkedMap
  cljs.core/IFn
  (-invoke
    ([m k] (.get m k nil))
    ([m k not-found] (.get m k not-found))))

(def m (goog.structs.LinkedMap.))
(.set m 34 :foo)
(println (m 34))
```

After I evaluated the expression `(println (m 34))`，I got a error message `#object[TypeError TypeError: app.main.m is not a function]`

**Comment by thheller** [comment]
The issue is that I added an optimization that emits direct function calls for JS types. Suppose you have `(:require ["react" :as r])` and `(r/createElement "div")`. Without the optimization CLJS would always emit

```
module$react.createElement.cljs$core$IFn$_invoke$arity$1 ? module$react.createElement.cljs$core$IFn$_invoke$arity$1("div") : module$react.createElement.call(null, "div")
```
with the optimization it just calls the fn
```
module$react.createElement("div")
```

This works much better with `:advanced`, produces less code overall and doesn't pay about 10% property-check tax for (almost) all function calls.

I did however not consider that someone would like to extend IFn on JS types as this then breaks for cases where the "type" of something is known. In the example `(def m ...)` will end up being tagged as `goog.structs.LinkedMap` which is a known JS type so the optimization kicks in.

I'm personally not willing to pay 10% extra overhead for function calls which is why I made the optimization. I may however put it behind a flag given that this optimization will likely never be accepted into CLJS directly because of this issue. `:fn-invoke-direct` sort of has the same issue which is why that has likely never become enabled by default.

It doesn't seem to be too common that people do this and only happens when extending IFn on JS types. Other protocols are not affected by this.
